### PR TITLE
Rate Limiting + API Framework

### LINKED ISSUE #1285: API rate limiting
Here's the plan I'm building for rate limiting, based on conversations on the gitter:
* Rate limiting user token is based on the user (for authenticated API access), or the IP address, for unauthenticated access
* Charge is tracked via apcu_inc (and will apcu_add to create an entry on failure)
* All API access will charge one unit by default. Available API charge and maybe reset time will be added as a response header.
* Will define API rate limits in constants, probably:
  * Authenticated: 5000 API calls per 5 minutes for authenticated users. (well in excess of what should be necessary)
  * Unauthenticated: 1000 API calls per 5 minutes
* Will add API calls to allow charging more than one unit for expensive operations.
* Will add alternate charge pools for operations we want to throttle heavily without impacting the performance of the website (e.g. random game API).

### PR DESCRIPTION
**Pull request by sgstair**
* Resolve #1285 - Add API that allows for rate limiting via `rateLimit_Charge(<cost>)` which returns true if the API is allowed to proceed. Further you can charge to alternate pools via `rateLimit_Charge(<cost>,"poolname",<initial treasury>, <number of seconds before charge is reset>)` - The default pool is `default` with 5000 units to charge and resets after 5 minutes.
* Add a new API framework to produce the kind of "nice" flat API definitions that have been kicked around in various threads (recently, mostly on gitter and #1287, though I remember this being brought up in an earlier issue I don't recall)
* Notifications API has been ported to use this new API definition method here and is the first API to include rate limiting. Stats is also in progress in other pending work

Simulator changes:
* Address #1261 by having the simulator automatically create default tags if they're not present.
* Simulator now uses the username as the password for new users, padded with "a" to 8 characters if necessary.
* Some user creation refactoring, and now the simulator cycles users creating 5 new users per event.

Future:
* I'm tracking work to change the status code returned when an API is rate limited, I realize there are better codes available but haven't looked it up yet.

### CODE REVIEW

File: src/shrub/src/core/ratelimit.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	}
+	
+	// Allow the call to succeed if the pool was not depleted before the charge was made.
+	return ($newvalue + $charge) > 0;
+}
+
+/// Check how much charge remains in a rate limiting pool
+/// @param Pool to use, or nothing for the default pool
+/// @retval integer count of charge remaining, or null if there is no pool currently.
+function rateLimit_RemainingCharge($poolname = null) {
```
Comment (povrazor) [inline src/shrub/src/core/ratelimit.php]:
```php
function rateLimit_RemainingCharge( $poolname = null ) {
```

File: src/shrub/src/core/ratelimit.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	$value = apcu_fetch($key);
+	if ( $value === false ) {
+		$value = null;
+	}
+	
+	return $value;
+}
+
+
+function rateLimit_Key($poolname) {
```
Comment (povrazor) [inline src/shrub/src/core/ratelimit.php]:
```php
function rateLimit_Key( $poolname ) {
```

**Comment by povrazor** [comment]
> About the only criticism I have is that the `API_CHARGE` part should maybe change in the future.
Rather than `API_CHARGE_1`, but rather we use constants for what class of API's are charged for the transaction.
> once we have them.
> That's a super easy change though.
> for now, `API_CHARGE_1` works as a placeholder.

> Probably do something clever, like `API_CHARGE_SEARCH | 1` to charge 1 unit of search to the call.
> or `API_COST_SEARCH`
> something like that

**Comment by sgstair** [comment]
That is certainly something we can do in the future, yeah

**Pull request merged**